python/CuTeDSL/cutlass/cute/core.py

```diff
@@ -4111,6 +4111,168 @@ def align_offset(offset, align):
         return (offset + (align - 1)) & ~(align - 1)
 
 
+class union(struct):
+    """
+    Decorator to abstract C union in Python DSL.
+
+    Similar to cute.struct, but lays out objects as a union:
+    - All objects start at offset 0
+    - The alignment is the maximum alignment of all objects
+    - The size is the maximum size of all objects
+
+    **Usage:**
+
+    .. code-block:: python
+
+        # Define a union with scalar int/float elements:
+        @cute.union
+        class value_union:
+            as_int : cutlass.Int32
+            as_float : cutlass.Float32
+
+
+        @cute.union
+        class data_union:
+            small : cutlass.Int16
+            medium : cutlass.Int32
+            large : cutlass.Int64
+
+
+        # Supports alignment for its elements:
+        @cute.union
+        class aligned_union:
+            a: cute.struct.Align[cutlass.Float32, 16]
+            b: cute.struct.Align[cutlass.Int32, 8]
+
+
+        # Statically get size and alignment:
+        size = data_union.__sizeof__()
+        align = data_union.__alignof__()
+
+        # Allocate and reference elements:
+        allocator = cutlass.utils.SmemAllocator()
+        value = allocator.allocate(data_union)
+
+        # Access union members (all at the same offset):
+        value.small ...
+        value.medium ...
+        value.large ...
+
+    :param cls: The union class with annotations.
+    :return: The decorated union class.
+    """
+
+    def __init__(self, cls):
```

Comment:
Probably also implement `__setitem__` for `union` to prevent user add new field after init?

Consider the following code

```python
@cute.union
class value_union:
    as_int : cutlass.Int32
    as_float : cutlass.Float32

def foo():
   vu = value_union()
   vu.as_some_craziness = myStruct() # This should be rejected

   vu.__sizeof__() # what should be the size now?
   
```

Reply:
added. struct doesn't have that either btw